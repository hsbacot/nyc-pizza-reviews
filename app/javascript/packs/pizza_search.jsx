// Run this example by adding <%= javascript_pack_tag 'hello_react' %> to the head of your layout file,
// like app/views/layouts/application.html.erb. All it does is render <div>Hello React</div> at the bottom
// of the page.

import React from 'react'
import ReactDOM from 'react-dom'
import PropTypes from 'prop-types'
import isEmpty from 'lodash/isEmpty'
import Review from './Review'

export class PizzaSearch extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      reviews: [],
      averageRating: '',
      restaurantSearch: '',
      restaurant: '',
      reviewCount: 5,
      loadingReviews: false
    }

    this.getReviews = this.getReviews.bind(this)
    this.handleInput = this.handleInput.bind(this)
    this.handleSearch = this.handleSearch.bind(this)
    this.handleNumericInput = this.handleNumericInput.bind(this)
  }

  getReviews(restaurant, count) {
    return fetch(`/pizzas/reviews?count=${count}&restaurant=${restaurant}`)
      .then(res => res.json())
  }

  handleInput(e) {
    this.setState({
      restaurantSearch: e.target.value
    })
  }

  handleSearch() {
    this.setState({
      restaurantSearch: '',
      reviewCount: 0,
      loadingReviews: true
    })
    this.getReviews(this.state.restaurantSearch, this.state.reviewCount).then(res => {
      this.setState({
        reviews: res.reviews,
        restaurant: res.restaurant,
        averageRating: res.average_rating,
        loadingReviews: false
      })
    })
  }

  handleNumericInput(e) {
    this.setState({
      reviewCount: e.target.value
    })
  }

  render() {
    const {
      averageRating,
      restaurantSearch,
      restaurant,
      reviewCount,
      reviews,
      loadingReviews
    } = this.state;

    const reviewItems = reviews.map(review => <Review copy={review.review_text} rating={review.rating} />)

    return(
      <div>
        <h2>NYC Pizza Reviews</h2>
        <p>Find pizzaria</p>
        <input value={restaurantSearch} onChange={this.handleInput} />
        <p>Number of reviews to fetch</p>
        <input type="number" value={reviewCount} onChange={this.handleNumericInput} />
        <button onClick={this.handleSearch}>Get Reviews</button>
        { restaurant && <h3>{restaurant} - average user rating: {averageRating}</h3>}
        { !isEmpty(reviews) && reviewItems}
        { loadingReviews && <p>loading...</p>}
      </div>
    )
  }
}

document.addEventListener('DOMContentLoaded', () => {
  ReactDOM.render(
    <PizzaSearch />,
    document.body.appendChild(document.createElement('div')),
  )
})
