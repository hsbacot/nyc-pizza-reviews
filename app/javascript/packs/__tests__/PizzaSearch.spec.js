import React from 'react'
import { PizzaSearch } from '../pizza_search'
// commented out due to webpack issues with enzyme dependencies and react updates
import { shallow } from 'enzyme'

const data = {"reviews":[{"review_text":"Awesome cheese pizza  - my go toEven though can not be your first choice, the caesar salad pizza is delicious ! Buffalo chicken pizza is also good","rating":4}],"average_rating":4.0,"restaurant":"Danny’s Pizzeria"}

describe('PizzaSearch', () => {
  it('should render PizzaSearch', () => {
    const component = shallow(<PizzaSearch />)
    expect(component).toMatchSnapshot()
  })

  it('should handle search input', () => {
    const component = shallow(<PizzaSearch />).instance()
    component.handleInput({ target: { value: 'dannys pizza' }})
    expect(component.state.restaurantSearch).toBe('dannys pizza')
  })

  it('should handle numerical input', () => {
    const component = shallow(<PizzaSearch />).instance()
    component.handleNumericInput({ target: { value: 3 }})
    expect(component.state.reviewCount).toBe(3)
  })

  it('should update the state to the response', async () => {
    const component = shallow(<PizzaSearch />).instance()
    const getReviewMock = jest.fn(() =>
      new Promise((resolve, reject) => resolve(data))
    );
    const setStateMock = jest.fn()
    component.getReviews = getReviewMock
    component.setState = setStateMock
    await component.handleSearch()
    expect(setStateMock.mock.calls.length).toBe(2);
    expect(setStateMock.mock.calls[1][0]).toHaveProperty('reviews', data.reviews)
    expect(setStateMock.mock.calls[1][0]).toHaveProperty('averageRating', 4)
    expect(setStateMock.mock.calls[1][0]).toHaveProperty('loadingReviews', false)
    expect(setStateMock.mock.calls[1][0]).toHaveProperty('restaurant')
  })

  it('should handle loading info', () => {
    const component = shallow(<PizzaSearch />).instance()
    component.setState({ loadingReviews: true })
    expect(component).toMatchSnapshot()
  })
})
