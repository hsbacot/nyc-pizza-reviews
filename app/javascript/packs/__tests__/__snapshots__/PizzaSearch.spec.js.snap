// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PizzaSearch should render PizzaSearch 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div>
    <h2>
        NYC Pizza Reviews
    </h2>
    <input
        onChange={[Function]}
        value=""
    />
    <input
        onChange={[Function]}
        type="number"
        value={5}
    />
    <button
        onClick={[Function]}
    >
        Get Reviews
    </button>
    
</div>,
  "nodes": Array [
    <div>
      <h2>
            NYC Pizza Reviews
      </h2>
      <input
            onChange={[Function]}
            value=""
      />
      <input
            onChange={[Function]}
            type="number"
            value={5}
      />
      <button
            onClick={[Function]}
      >
            Get Reviews
      </button>
      
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <PizzaSearch />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": PizzaSearch {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "getReviews": [Function],
        "handleInput": [Function],
        "handleNumericInput": [Function],
        "handleSearch": [Function],
        "props": Object {},
        "refs": Object {},
        "state": Object {
          "averageRating": "",
          "restaurant": "",
          "restaurantSearch": "",
          "reviewCount": 5,
          "reviews": Array [],
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div>
          <h2>
                    NYC Pizza Reviews
          </h2>
          <input
                    onChange={[Function]}
                    value=""
          />
          <input
                    onChange={[Function]}
                    type="number"
                    value={5}
          />
          <button
                    onClick={[Function]}
          >
                    Get Reviews
          </button>
          
</div>,
        "_debugID": 2,
        "_renderedOutput": <div>
          <h2>
                    NYC Pizza Reviews
          </h2>
          <input
                    onChange={[Function]}
                    value=""
          />
          <input
                    onChange={[Function]}
                    type="number"
                    value={5}
          />
          <button
                    onClick={[Function]}
          >
                    Get Reviews
          </button>
          
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <PizzaSearch />,
}
`;
